src/coloring/backtracking_coloring.jl

```diff
@@ -0,0 +1,238 @@
+using LightGraphs
+
+"""
+    color_graph(g::LightGraphs, ::BacktrackingColor)
+
+Returns a tight, distance-1 coloring of graph g
+using the minimum number of colors possible (i.e.
+the chromatic number of graph, χ(g))
+"""
+function color_graph(g::LightGraphs.AbstractGraph, ::BacktrackingColor)
+    v = nv(g)
+
+    #A is list of vertices in non-increasing order of degree
+    A = sort_by_degree(g)
+
+    #F is the coloring of vertices, 0 means uncolored
+    #Fopt is the optimal coloring of the graph
+    F = zeros(Int, v)
+    Fopt= zeros(Int, v)
+
+    start = 1
+
+    #optimal color number
+    opt = v + 1
+
+    #current vertex to be colored
+    x = A[1]
+
+    #colors[j] = number of colors in A[0]...A[j]
+    #assume colors[0] = 1
+    colors = zeros(Int, v)
+
+    #set of free colors
+    U = zeros(Int, 0)
+    push!(U, 1)
+
+    #set of free colors of x
+    freeColors = [Vector{Int}() for _ in 1:v]
+    freeColors[x] = copy(U)
+
+    while (start >= 1)
+
+        back = false
+        for i = start:v
+            if i > start
+                x = uncolored_vertex_of_maximal_degree(A,F)
+                U = free_colors(x, A, colors, F, g, opt)
+                sort!(U)
+            end
+            if length(U) > 0
+                k = U[1]
+                F[x] = k
+                cp = F[x]
+                deleteat!(U, 1)
+                freeColors[x] = copy(U)
+                if i==1
+                    l = 0
+                else
+                    l = colors[i-1]
+                end
+                colors[i] = max(k, l)
+            else
+                start = i-1
+                back = true
+                break
+            end
+        end
+
+        if back
+            if start >= 1
+                x = A[start]
+                F[x] = 0
+                U = freeColors[x]
+            end
+        else
+            Fopt = copy(F)
+            opt = colors[v-1]
+            i = least_index(F,A,opt)
+            start = i-1
+            if start < 1
+                break
+            end
+
+            #uncolor all vertices A[i] with i >= start
+            uncolor_all!(F, A, start)
+
+            for i = 1:start+1
+                x = A[i]
+                U = freeColors[x]
+
+                #remove colors >= opt from U
+                U = remove_higher_colors(U, opt)
+                freeColors[x] = copy(U)
+            end
+        end
+
+    end
+
+    Fopt
```

Comment:
```suggestion
    return Fopt
```
You can remove the empty lines above and below

Reply:
Any specific reason why `return x` is better than `x`?

Reply:
For long form functions, it is felt by many that explicit `return` statements are better style: clearer intention, less likely that something goes wrong in a subsequent refactor.  If I recall correctly, I saw Stefan commenting somewhere, that he regrets that `return` is not mandatory for long-form functions.

Reply:
Thanks for the explanation.